track_rotated

rockb1017 opened the pull request:
fixes #85 
implementing https://github.com/open-telemetry/opentelemetry-log-collection/issues/85#issuecomment-852280642

in addition 
* i removed unnecessary fingerprint comparison for copytrunc cases because it is repetitive. 
* I modified file batching test case (2nd part of test) because with this new approach, the way it handles already-discovered-recently-appended log files differently.

djaglowski commented on operator/builtin/input/file/file_test.go:
```diff
@@ -469,24 +470,6 @@ func TestFileBatching(t *testing.T) {
 	}
 
 	require.ElementsMatch(t, expectedMessages, actualMessages)
-
-	// Write more logs to each file so we can validate that all files are still known
-	for i, temp := range temps {
-		for j := 0; j < linesPerFile; j++ {
-			message := fmt.Sprintf("%s %d %d", stringWithLength(20), i, j)
-			temp.WriteString(message + "\n")
-			expectedMessages = append(expectedMessages, message)
-		}
-	}
-
-	for b := 0; b < expectedBatches; b++ {
-		// poll once so we can validate that files were batched
-		operator.poll(context.Background())
-		actualMessages = append(actualMessages, waitForN(t, logReceived, expectedLinesPerBatch)...)
-		expectNoMessagesUntil(t, logReceived, 10*time.Millisecond)
-	}
-
-	require.ElementsMatch(t, expectedMessages, actualMessages)
```
I think these changes have invalidated the test.

Tinkering with the removed code shows me that the first `for` loop does not cause an issue. In other words, we can still "Write more logs to each file ...".

However, any additional call to `operator.poll` will never return.

I think this needs to be understood carefully, and that it likely indicates a problem with this implementation.

rockb1017 replied:
It was because the mock channel created has limited size. I doubled it and it works. at the first poll cycle of 2nd for loop, it tails 20 files that have data added. so the number of events in that cycle is double of rest of the cycles. Anyway, The MaxConcurrentFiles is respected.

djaglowski reviewed (commented):
I really like how simple this implementation is. It's much improved from the past version.

There is at least one significant issue that we needs to be addressed, relating to the modified test.

I have the following additional requests:
1. Will you please add a note to the documentation, which describes the interaction between `max_concurrent_files` and files that are rotated out of the matching pattern.
2. Will you please rebase onto `main`, so we can run the new benchmarks before/after this change, and validate that performance changes are acceptable.

rockb1017 commented:
I added closing all readers step to `Stop()` function. and removed fingerprint comparison again because it is not needed.

rockb1017 commented:
hmm for windows, it fails `TestMultiFileParallel_LiveFiles` test. it consumed file1 and file4 twice. 
I can't reproduce it. any suggestion? 
Could you rerun just in case?

rockb1017 commented:
hmm but it is a different test that failed. `TestRotation/NoDeletion/MoveCreateSequential`

rockb1017 commented:
Did increasing poll interval made it fail or was it because test failed before it got to run `TestRotation/NoDeletion/MoveCreateSequential` test?

rockb1017 commented:
I couldn't trace back and understand why it failed on windows. So i spun up a windows server on EC2 and ran test. 
```
Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test -v
=== RUN   TestUnmarshal
=== RUN   TestUnmarshal/default
=== RUN   TestUnmarshal/extra_field
=== RUN   TestUnmarshal/id_custom
=== RUN   TestUnmarshal/include_one
=== RUN   TestUnmarshal/include_multi
=== RUN   TestUnmarshal/include_glob
=== RUN   TestUnmarshal/include_glob_double_asterisk
=== RUN   TestUnmarshal/include_glob_double_asterisk_nested
=== RUN   TestUnmarshal/include_glob_double_asterisk_prefix
=== RUN   TestUnmarshal/include_inline
=== RUN   TestUnmarshal/include_invalid
=== RUN   TestUnmarshal/exclude_one
=== RUN   TestUnmarshal/exclude_multi
=== RUN   TestUnmarshal/exclude_glob
=== RUN   TestUnmarshal/exclude_glob_double_asterisk
=== RUN   TestUnmarshal/exclude_glob_double_asterisk_nested
=== RUN   TestUnmarshal/exclude_glob_double_asterisk_prefix
=== RUN   TestUnmarshal/exclude_inline
=== RUN   TestUnmarshal/exclude_invalid
=== RUN   TestUnmarshal/poll_interval_no_units
=== RUN   TestUnmarshal/poll_interval_1s
=== RUN   TestUnmarshal/poll_interval_1ms
=== RUN   TestUnmarshal/poll_interval_1000ms
=== RUN   TestUnmarshal/fingerprint_size_no_units
=== RUN   TestUnmarshal/fingerprint_size_1kb_lower
=== RUN   TestUnmarshal/fingerprint_size_1KB
=== RUN   TestUnmarshal/fingerprint_size_1kib_lower
=== RUN   TestUnmarshal/fingerprint_size_1KiB
=== RUN   TestUnmarshal/fingerprint_size_float
=== RUN   TestUnmarshal/include_file_name_lower
=== RUN   TestUnmarshal/include_file_name_upper
=== RUN   TestUnmarshal/include_file_name_on
=== RUN   TestUnmarshal/include_file_name_yes
=== RUN   TestUnmarshal/include_file_path_lower
=== RUN   TestUnmarshal/include_file_path_upper
=== RUN   TestUnmarshal/include_file_path_on
=== RUN   TestUnmarshal/include_file_path_yes
=== RUN   TestUnmarshal/include_file_path_off
=== RUN   TestUnmarshal/include_file_path_no
=== RUN   TestUnmarshal/include_file_path_nonbool
=== RUN   TestUnmarshal/multiline_line_start_string
=== RUN   TestUnmarshal/multiline_line_start_special
=== RUN   TestUnmarshal/multiline_line_end_string
=== RUN   TestUnmarshal/multiline_line_end_special
=== RUN   TestUnmarshal/multiline_random
=== RUN   TestUnmarshal/start_at_string
=== RUN   TestUnmarshal/max_concurrent_large
=== RUN   TestUnmarshal/max_log_size_mib_lower
=== RUN   TestUnmarshal/max_log_size_mib_upper
=== RUN   TestUnmarshal/max_log_size_mb_upper
=== RUN   TestUnmarshal/max_log_size_mb_lower
=== RUN   TestUnmarshal/encoding_lower
=== RUN   TestUnmarshal/encoding_upper
--- PASS: TestUnmarshal (0.02s)
    --- PASS: TestUnmarshal/default (0.00s)
    --- PASS: TestUnmarshal/extra_field (0.00s)
    --- PASS: TestUnmarshal/id_custom (0.00s)
    --- PASS: TestUnmarshal/include_one (0.00s)
    --- PASS: TestUnmarshal/include_multi (0.00s)
    --- PASS: TestUnmarshal/include_glob (0.00s)
    --- PASS: TestUnmarshal/include_glob_double_asterisk (0.00s)
    --- PASS: TestUnmarshal/include_glob_double_asterisk_nested (0.00s)
    --- PASS: TestUnmarshal/include_glob_double_asterisk_prefix (0.00s)
    --- PASS: TestUnmarshal/include_inline (0.00s)
    --- PASS: TestUnmarshal/include_invalid (0.00s)
    --- PASS: TestUnmarshal/exclude_one (0.00s)
    --- PASS: TestUnmarshal/exclude_multi (0.00s)
    --- PASS: TestUnmarshal/exclude_glob (0.00s)
    --- PASS: TestUnmarshal/exclude_glob_double_asterisk (0.00s)
    --- PASS: TestUnmarshal/exclude_glob_double_asterisk_nested (0.00s)
    --- PASS: TestUnmarshal/exclude_glob_double_asterisk_prefix (0.00s)
    --- PASS: TestUnmarshal/exclude_inline (0.00s)
    --- PASS: TestUnmarshal/exclude_invalid (0.00s)
    --- PASS: TestUnmarshal/poll_interval_no_units (0.00s)
    --- PASS: TestUnmarshal/poll_interval_1s (0.00s)
    --- PASS: TestUnmarshal/poll_interval_1ms (0.00s)
    --- PASS: TestUnmarshal/poll_interval_1000ms (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_no_units (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_1kb_lower (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_1KB (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_1kib_lower (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_1KiB (0.00s)
    --- PASS: TestUnmarshal/fingerprint_size_float (0.00s)
    --- PASS: TestUnmarshal/include_file_name_lower (0.00s)
    --- PASS: TestUnmarshal/include_file_name_upper (0.00s)
    --- PASS: TestUnmarshal/include_file_name_on (0.00s)
    --- PASS: TestUnmarshal/include_file_name_yes (0.00s)
    --- PASS: TestUnmarshal/include_file_path_lower (0.00s)
    --- PASS: TestUnmarshal/include_file_path_upper (0.00s)
    --- PASS: TestUnmarshal/include_file_path_on (0.00s)
    --- PASS: TestUnmarshal/include_file_path_yes (0.00s)
    --- PASS: TestUnmarshal/include_file_path_off (0.00s)
    --- PASS: TestUnmarshal/include_file_path_no (0.00s)
    --- PASS: TestUnmarshal/include_file_path_nonbool (0.00s)
    --- PASS: TestUnmarshal/multiline_line_start_string (0.00s)
    --- PASS: TestUnmarshal/multiline_line_start_special (0.00s)
    --- PASS: TestUnmarshal/multiline_line_end_string (0.00s)
    --- PASS: TestUnmarshal/multiline_line_end_special (0.00s)
    --- PASS: TestUnmarshal/multiline_random (0.01s)
    --- PASS: TestUnmarshal/start_at_string (0.00s)
    --- PASS: TestUnmarshal/max_concurrent_large (0.00s)
    --- PASS: TestUnmarshal/max_log_size_mib_lower (0.00s)
    --- PASS: TestUnmarshal/max_log_size_mib_upper (0.00s)
    --- PASS: TestUnmarshal/max_log_size_mb_upper (0.00s)
    --- PASS: TestUnmarshal/max_log_size_mb_lower (0.00s)
    --- PASS: TestUnmarshal/encoding_lower (0.00s)
    --- PASS: TestUnmarshal/encoding_upper (0.00s)
=== RUN   TestBuild
=== PAUSE TestBuild
=== RUN   TestMapStructureDecodeConfigWithHook
--- PASS: TestMapStructureDecodeConfigWithHook (0.00s)
=== RUN   TestMapStructureDecodeConfig
--- PASS: TestMapStructureDecodeConfig (0.00s)
=== RUN   TestCleanStop
=== PAUSE TestCleanStop
=== RUN   TestAddFileFields
=== PAUSE TestAddFileFields
=== RUN   TestReadExistingLogs
=== PAUSE TestReadExistingLogs
=== RUN   TestReadNewLogs
=== PAUSE TestReadNewLogs
=== RUN   TestReadExistingAndNewLogs
=== PAUSE TestReadExistingAndNewLogs
=== RUN   TestStartAtEnd
=== PAUSE TestStartAtEnd
=== RUN   TestStartAtEndNewFile
=== PAUSE TestStartAtEndNewFile
=== RUN   TestNoNewline
=== PAUSE TestNoNewline
=== RUN   TestSkipEmpty
=== PAUSE TestSkipEmpty
=== RUN   TestSplitWrite
=== PAUSE TestSplitWrite
=== RUN   TestDecodeBufferIsResized
=== PAUSE TestDecodeBufferIsResized
=== RUN   TestMultiFileSimple
=== PAUSE TestMultiFileSimple
=== RUN   TestMultiFileParallel_PreloadedFiles
=== PAUSE TestMultiFileParallel_PreloadedFiles
=== RUN   TestMultiFileParallel_LiveFiles
=== PAUSE TestMultiFileParallel_LiveFiles
=== RUN   TestOffsetsAfterRestart
=== PAUSE TestOffsetsAfterRestart
=== RUN   TestOffsetsAfterRestart_BigFiles
=== PAUSE TestOffsetsAfterRestart_BigFiles
=== RUN   TestOffsetsAfterRestart_BigFilesWrittenWhileOff
=== PAUSE TestOffsetsAfterRestart_BigFilesWrittenWhileOff
=== RUN   TestManyLogsDelivered
=== PAUSE TestManyLogsDelivered
=== RUN   TestFileBatching
=== PAUSE TestFileBatching
=== RUN   TestFileReader_FingerprintUpdated
=== PAUSE TestFileReader_FingerprintUpdated
=== RUN   TestEncodings
=== PAUSE TestEncodings
=== RUN   TestExclude
--- PASS: TestExclude (0.00s)
=== RUN   TestExcludeEmpty
--- PASS: TestExcludeEmpty (0.00s)
=== RUN   TestExcludeMany
--- PASS: TestExcludeMany (0.01s)
=== RUN   TestExcludeDuplicates
--- PASS: TestExcludeDuplicates (0.00s)
=== RUN   TestNewFingerprintDoesNotModifyOffset
--- PASS: TestNewFingerprintDoesNotModifyOffset (0.00s)
=== RUN   TestNewFingerprint
=== RUN   TestNewFingerprint/defaultExactFileSize
=== PAUSE TestNewFingerprint/defaultExactFileSize
=== RUN   TestNewFingerprint/defaultWithFileHalfOfFingerprint
=== PAUSE TestNewFingerprint/defaultWithFileHalfOfFingerprint
=== RUN   TestNewFingerprint/defaultWithFileTwiceFingerprint
=== PAUSE TestNewFingerprint/defaultWithFileTwiceFingerprint
=== RUN   TestNewFingerprint/minFingerprintExactFileSize
=== PAUSE TestNewFingerprint/minFingerprintExactFileSize
=== RUN   TestNewFingerprint/minFingerprintWithSmallerFileSize
=== PAUSE TestNewFingerprint/minFingerprintWithSmallerFileSize
=== RUN   TestNewFingerprint/minFingerprintWithLargerFileSize
=== PAUSE TestNewFingerprint/minFingerprintWithLargerFileSize
=== RUN   TestNewFingerprint/largeFingerprintSmallFile
=== PAUSE TestNewFingerprint/largeFingerprintSmallFile
=== RUN   TestNewFingerprint/largeFingerprintLargeFile
=== PAUSE TestNewFingerprint/largeFingerprintLargeFile
=== CONT  TestNewFingerprint/defaultExactFileSize
=== CONT  TestNewFingerprint/minFingerprintWithLargerFileSize
=== CONT  TestNewFingerprint/largeFingerprintLargeFile
=== CONT  TestNewFingerprint/largeFingerprintSmallFile
=== CONT  TestNewFingerprint/minFingerprintExactFileSize
=== CONT  TestNewFingerprint/defaultWithFileTwiceFingerprint
=== CONT  TestNewFingerprint/minFingerprintWithSmallerFileSize
=== CONT  TestNewFingerprint/defaultWithFileHalfOfFingerprint
--- PASS: TestNewFingerprint (0.00s)
    --- PASS: TestNewFingerprint/defaultExactFileSize (0.15s)
    --- PASS: TestNewFingerprint/largeFingerprintLargeFile (0.17s)
    --- PASS: TestNewFingerprint/minFingerprintWithLargerFileSize (0.17s)
    --- PASS: TestNewFingerprint/largeFingerprintSmallFile (0.18s)
    --- PASS: TestNewFingerprint/minFingerprintExactFileSize (0.18s)
    --- PASS: TestNewFingerprint/defaultWithFileHalfOfFingerprint (0.18s)
    --- PASS: TestNewFingerprint/minFingerprintWithSmallerFileSize (0.18s)
    --- PASS: TestNewFingerprint/defaultWithFileTwiceFingerprint (0.18s)
=== RUN   TestFingerprintCopy
=== PAUSE TestFingerprintCopy
=== RUN   TestFingerprintStartsWith
=== RUN   TestFingerprintStartsWith/same
=== PAUSE TestFingerprintStartsWith/same
=== RUN   TestFingerprintStartsWith/aStartsWithB
=== PAUSE TestFingerprintStartsWith/aStartsWithB
=== RUN   TestFingerprintStartsWith/bStartsWithA
=== PAUSE TestFingerprintStartsWith/bStartsWithA
=== RUN   TestFingerprintStartsWith/neither
=== PAUSE TestFingerprintStartsWith/neither
=== CONT  TestFingerprintStartsWith/same
=== CONT  TestFingerprintStartsWith/bStartsWithA
=== CONT  TestFingerprintStartsWith/neither
=== CONT  TestFingerprintStartsWith/aStartsWithB
--- PASS: TestFingerprintStartsWith (0.00s)
    --- PASS: TestFingerprintStartsWith/same (0.00s)
    --- PASS: TestFingerprintStartsWith/bStartsWithA (0.00s)
    --- PASS: TestFingerprintStartsWith/neither (0.00s)
    --- PASS: TestFingerprintStartsWith/aStartsWithB (0.00s)
=== RUN   TestFingerprintStartsWith_FromFile
--- PASS: TestFingerprintStartsWith_FromFile (1.08s)
=== RUN   TestMultiFileRotate
=== PAUSE TestMultiFileRotate
=== RUN   TestMultiFileRotateSlow
    rotation_test.go:88:
--- SKIP: TestMultiFileRotateSlow (0.00s)
=== RUN   TestMultiCopyTruncateSlow
--- PASS: TestMultiCopyTruncateSlow (2.42s)
=== RUN   TestRotation
=== RUN   TestRotation/NoRotation/MoveCreateTimestamped
=== RUN   TestRotation/NoRotation/MoveCreateSequential
=== RUN   TestRotation/NoRotation/CopyTruncateTimestamped
=== RUN   TestRotation/NoRotation/CopyTruncateSequential
=== RUN   TestRotation/NoDeletion/MoveCreateTimestamped
=== RUN   TestRotation/NoDeletion/MoveCreateSequential
=== RUN   TestRotation/NoDeletion/CopyTruncateTimestamped
=== RUN   TestRotation/NoDeletion/CopyTruncateSequential
=== RUN   TestRotation/Deletion/MoveCreateTimestamped
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion
=== RUN   TestRotation/Deletion/MoveCreateSequential
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion
=== RUN   TestRotation/Deletion/CopyTruncateTimestamped
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion
=== RUN   TestRotation/Deletion/CopyTruncateSequential
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion
=== RUN   TestRotation/Deletion/ExceedFingerprint/MoveCreateTimestamped
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion/ExceedFingerprint
=== RUN   TestRotation/Deletion/ExceedFingerprint/MoveCreateSequential
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion/ExceedFingerprint
=== RUN   TestRotation/Deletion/ExceedFingerprint/CopyTruncateTimestamped
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion/ExceedFingerprint
=== RUN   TestRotation/Deletion/ExceedFingerprint/CopyTruncateSequential
    rotation_test.go:277: Potentially unstable ephemerality expectation for test: Deletion/ExceedFingerprint
--- PASS: TestRotation (22.00s)
    --- PASS: TestRotation/NoRotation/MoveCreateTimestamped (0.32s)
    --- PASS: TestRotation/NoRotation/MoveCreateSequential (0.34s)
    --- PASS: TestRotation/NoRotation/CopyTruncateTimestamped (0.33s)
    --- PASS: TestRotation/NoRotation/CopyTruncateSequential (0.31s)
    --- PASS: TestRotation/NoDeletion/MoveCreateTimestamped (0.48s)
    --- PASS: TestRotation/NoDeletion/MoveCreateSequential (0.40s)
    --- PASS: TestRotation/NoDeletion/CopyTruncateTimestamped (0.46s)
    --- PASS: TestRotation/NoDeletion/CopyTruncateSequential (0.48s)
    --- PASS: TestRotation/Deletion/MoveCreateTimestamped (0.65s)
    --- PASS: TestRotation/Deletion/MoveCreateSequential (0.60s)
    --- PASS: TestRotation/Deletion/CopyTruncateTimestamped (0.60s)
    --- PASS: TestRotation/Deletion/CopyTruncateSequential (0.59s)
    --- PASS: TestRotation/Deletion/ExceedFingerprint/MoveCreateTimestamped (4.17s)
    --- PASS: TestRotation/Deletion/ExceedFingerprint/MoveCreateSequential (4.19s)
    --- PASS: TestRotation/Deletion/ExceedFingerprint/CopyTruncateTimestamped (4.04s)
    --- PASS: TestRotation/Deletion/ExceedFingerprint/CopyTruncateSequential (4.04s)
=== RUN   TestMoveFile
    rotation_test.go:334: Moving files while open is unsupported on Windows
--- SKIP: TestMoveFile (0.00s)
=== RUN   TestTrackMovedAwayFiles
    rotation_test.go:360: Moving files while open is unsupported on Windows
--- SKIP: TestTrackMovedAwayFiles (0.00s)
=== RUN   TestTruncateThenWrite
=== PAUSE TestTruncateThenWrite
=== RUN   TestCopyTruncateWriteBoth
=== PAUSE TestCopyTruncateWriteBoth
=== RUN   TestFileMovedWhileOff_BigFiles
=== PAUSE TestFileMovedWhileOff_BigFiles
=== CONT  TestBuild
=== CONT  TestEncodings
=== CONT  TestStartAtEndNewFile
=== RUN   TestEncodings/Nop
=== PAUSE TestEncodings/Nop
=== RUN   TestEncodings/InvalidUTFReplacement
=== PAUSE TestEncodings/InvalidUTFReplacement
=== RUN   TestBuild/Basic
=== CONT  TestMultiFileParallel_PreloadedFiles
=== CONT  TestMultiFileSimple
=== CONT  TestDecodeBufferIsResized
=== CONT  TestSplitWrite
=== CONT  TestCopyTruncateWriteBoth
=== RUN   TestEncodings/ValidUTF8
=== PAUSE TestBuild/Basic
=== PAUSE TestEncodings/ValidUTF8
=== RUN   TestBuild/BadIncludeGlob
=== PAUSE TestBuild/BadIncludeGlob
=== RUN   TestEncodings/ChineseCharacter
=== RUN   TestBuild/BadExcludeGlob
=== PAUSE TestBuild/BadExcludeGlob
=== PAUSE TestEncodings/ChineseCharacter
=== RUN   TestEncodings/SmileyFaceUTF16
=== RUN   TestBuild/MultilineConfiguredStartAndEndPatterns
=== PAUSE TestBuild/MultilineConfiguredStartAndEndPatterns
=== PAUSE TestEncodings/SmileyFaceUTF16
=== RUN   TestBuild/MultilineConfiguredStartPattern
=== PAUSE TestBuild/MultilineConfiguredStartPattern
=== RUN   TestEncodings/SmileyFaceNewlineUTF16
=== RUN   TestBuild/MultilineConfiguredEndPattern
=== PAUSE TestEncodings/SmileyFaceNewlineUTF16
=== PAUSE TestBuild/MultilineConfiguredEndPattern
=== RUN   TestEncodings/SmileyFaceNewlineUTF16LE
=== RUN   TestBuild/InvalidEncoding
=== PAUSE TestEncodings/SmileyFaceNewlineUTF16LE
=== PAUSE TestBuild/InvalidEncoding
=== RUN   TestBuild/LineStartAndEnd
=== PAUSE TestBuild/LineStartAndEnd
=== RUN   TestEncodings/ChineseCharacterBig5
=== PAUSE TestEncodings/ChineseCharacterBig5
=== RUN   TestBuild/NoLineStartOrEnd
=== PAUSE TestBuild/NoLineStartOrEnd
=== CONT  TestSkipEmpty
=== RUN   TestBuild/InvalidLineStartRegex
=== PAUSE TestBuild/InvalidLineStartRegex
=== RUN   TestBuild/InvalidLineEndRegex
=== PAUSE TestBuild/InvalidLineEndRegex
=== CONT  TestNoNewline
    file_test.go:174:
--- SKIP: TestNoNewline (0.00s)
=== CONT  TestReadNewLogs
--- PASS: TestStartAtEndNewFile (0.01s)
=== CONT  TestStartAtEnd
--- PASS: TestReadNewLogs (0.01s)
=== CONT  TestReadExistingAndNewLogs
--- PASS: TestSplitWrite (0.01s)
=== CONT  TestAddFileFields
--- PASS: TestReadExistingAndNewLogs (0.00s)
=== CONT  TestReadExistingLogs
--- PASS: TestStartAtEnd (0.21s)
=== CONT  TestFileMovedWhileOff_BigFiles
--- PASS: TestSkipEmpty (0.21s)
=== CONT  TestCleanStop
    file_test.go:36: Skipping due to goroutine leak in opencensus.
        See this issue for details: https://github.com/census-instrumentation/opencensus-go/issues/1191#issuecomment-610440163
=== CONT  TestManyLogsDelivered
--- SKIP: TestCleanStop (0.00s)
=== CONT  TestFileReader_FingerprintUpdated
--- PASS: TestAddFileFields (0.20s)
=== CONT  TestFileBatching
--- PASS: TestDecodeBufferIsResized (0.21s)
--- PASS: TestFileReader_FingerprintUpdated (0.00s)
=== CONT  TestOffsetsAfterRestart_BigFiles
--- PASS: TestReadExistingLogs (0.21s)
=== CONT  TestOffsetsAfterRestart
--- PASS: TestFileBatching (0.28s)
=== CONT  TestOffsetsAfterRestart_BigFilesWrittenWhileOff
=== CONT  TestMultiFileParallel_LiveFiles
--- PASS: TestManyLogsDelivered (0.41s)
--- PASS: TestOffsetsAfterRestart_BigFiles (0.41s)
=== CONT  TestMultiFileRotate
--- PASS: TestFileMovedWhileOff_BigFiles (0.41s)
=== CONT  TestTruncateThenWrite
--- PASS: TestOffsetsAfterRestart (0.40s)
=== CONT  TestFingerprintCopy
--- PASS: TestFingerprintCopy (0.00s)
=== CONT  TestEncodings/Nop
--- PASS: TestTruncateThenWrite (0.21s)
=== CONT  TestEncodings/SmileyFaceNewlineUTF16
=== CONT  TestEncodings/SmileyFaceUTF16
--- PASS: TestOffsetsAfterRestart_BigFilesWrittenWhileOff (0.42s)
=== CONT  TestEncodings/ChineseCharacter
--- PASS: TestCopyTruncateWriteBoth (1.01s)
=== CONT  TestEncodings/ValidUTF8
=== CONT  TestEncodings/InvalidUTFReplacement
=== CONT  TestEncodings/ChineseCharacterBig5
=== CONT  TestEncodings/SmileyFaceNewlineUTF16LE
=== CONT  TestBuild/Basic
--- PASS: TestMultiFileSimple (1.21s)
=== CONT  TestBuild/InvalidEncoding
=== CONT  TestBuild/InvalidLineEndRegex
=== CONT  TestBuild/InvalidLineStartRegex
=== CONT  TestBuild/NoLineStartOrEnd
=== CONT  TestBuild/LineStartAndEnd
=== CONT  TestBuild/MultilineConfiguredStartAndEndPatterns
=== CONT  TestBuild/MultilineConfiguredEndPattern
=== CONT  TestBuild/MultilineConfiguredStartPattern
=== CONT  TestBuild/BadExcludeGlob
=== CONT  TestBuild/BadIncludeGlob
--- PASS: TestBuild (0.00s)
    --- PASS: TestBuild/InvalidEncoding (0.00s)
    --- PASS: TestBuild/InvalidLineEndRegex (0.00s)
    --- PASS: TestBuild/Basic (0.00s)
    --- PASS: TestBuild/InvalidLineStartRegex (0.00s)
    --- PASS: TestBuild/LineStartAndEnd (0.00s)
    --- PASS: TestBuild/MultilineConfiguredStartAndEndPatterns (0.00s)
    --- PASS: TestBuild/NoLineStartOrEnd (0.00s)
    --- PASS: TestBuild/MultilineConfiguredStartPattern (0.00s)
    --- PASS: TestBuild/MultilineConfiguredEndPattern (0.00s)
    --- PASS: TestBuild/BadExcludeGlob (0.00s)
    --- PASS: TestBuild/BadIncludeGlob (0.00s)
--- PASS: TestMultiFileParallel_PreloadedFiles (1.25s)
--- PASS: TestEncodings (0.00s)
    --- PASS: TestEncodings/Nop (0.20s)
    --- PASS: TestEncodings/SmileyFaceNewlineUTF16 (0.20s)
    --- PASS: TestEncodings/SmileyFaceUTF16 (0.21s)
    --- PASS: TestEncodings/ChineseCharacter (0.22s)
    --- PASS: TestEncodings/ValidUTF8 (0.20s)
    --- PASS: TestEncodings/InvalidUTFReplacement (0.20s)
    --- PASS: TestEncodings/ChineseCharacterBig5 (0.21s)
    --- PASS: TestEncodings/SmileyFaceNewlineUTF16LE (0.20s)
--- PASS: TestMultiFileRotate (1.21s)
--- PASS: TestMultiFileParallel_LiveFiles (1.24s)
PASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      27.607s

```

everything passes. 
i kept running just in case. 
```
Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test
PASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      27.225s

Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test
gPASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      20.838s

Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test
PASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      20.087s

Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test
PASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      27.322s

Administrator@EC2AMAZ-K6DFUOR MINGW64 ~/opentelemetry-log-collection/operator/builtin/input/file (track_rotated)
$ go test
PASS
ok      github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file      26.866s
```
it passes without failing. 
Any clue on why the test is flaky on github actions?

rockb1017 commented:
at this point,I think i am just fixing the flaky test on windows machine because I don't have any failing cases on my windows. 

1. I have rebased it.
2. I have increase timeouts for unit test in the attempt to fix it.

djaglowski commented:
> at this point,I think i am just fixing the flaky test on windows machine because I don't have any failing cases on my windows.
> 
> 1. I have rebased it.
> 2. I have increase timeouts for unit test in the attempt to fix it.

I think we should not look at this as a matter of flaky tests. These tests have no history of being flaky, to my knowledge. Tuning timeouts is sometimes necessary at a macro-level, but our tests should pass with timeouts that span a reasonable range.

Instead, let's assume that these changes may have introduced an issue that is causing the problem. 

We see the following error on several failing windows tests: `The process cannot access the file because it is being used by another process` 

The nature of this PR is to change the way in which we are "using" files, so it seems likely that these changes are directly related.

tigrannajaryan commented on operator/builtin/input/file/rotation_test.go:
```diff
@@ -355,6 +355,42 @@ func TestMoveFile(t *testing.T) {
 
+func TestTrackMovedAwayFiles(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("Moving files while open is unsupported on Windows")
```
I think opening with with FILE_SHARE_DELETE flag should make this possible. See https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea?redirectedfrom=MSDN

rockb1017 commented:
yes failing test after rebase is a little different. so i ran it on my windows machine, and yay i can reproduce them. 
I added `defer operator.Stop()` in tests. test itself tries to delete the file after testing, but operator has them open. that was the issue. after I added it, it passed on my windows. let's see if what github action says.

rockb1017 commented:
it passed. feel free to re run pipeline just to check.

djaglowski commented:
I tried running benchmarks on this and found a problem. Some clarification is needed.

1. #157 introduced a noisy log message, which I think can be ignored. I will open a PR shortly. In the meantime you can search for "Problem closing reader" and comment out the log.
2. This benchmark is apparently demonstrating a problem that unit tests are not. I will try to convert this into a unit test so that we can catch this more easily.

The problem is that the benchmark never completes. It hangs on the `MaxConcurrent` case.

The good news is that, for the 3 cases that do complete, we see almost identical performance. Nice job.

On main:
```
> make bench
go test -benchmem -run=^$ -bench ^* ./...
...
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file
BenchmarkFileInput/Single-8               652237              2193 ns/op             688 B/op         10 allocs/op
BenchmarkFileInput/Glob-8                 111085             10936 ns/op            2762 B/op         42 allocs/op
BenchmarkFileInput/MultiGlob-8            113517             10727 ns/op            2762 B/op         42 allocs/op
BenchmarkFileInput/MaxConcurrent-8         41650             24309 ns/op            2757 B/op         42 allocs/op
BenchmarkFileInput/FngrPrntLarge-8        485581              2089 ns/op             689 B/op         10 allocs/op
BenchmarkFileInput/FngrPrntSmall-8        487039              2088 ns/op             688 B/op         10 allocs/op
```

```
> make bench
go test -benchmem -run=^$ -bench ^* ./...
...
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file
BenchmarkFileInput/Single-8               473388              2140 ns/op             688 B/op         10 allocs/op
BenchmarkFileInput/Glob-8                 116604             10445 ns/op            2761 B/op         42 allocs/op
BenchmarkFileInput/MultiGlob-8            114943             10600 ns/op            2762 B/op         42 allocs/op
```

djaglowski commented:
#174 and #176 have been merged. 

174 will take care of the noisy "problem closing reader" logs, but you'll have to determine whether a similar change makes sense on this branch.

176 adds benchmarks to CI, which should demonstrate the problem. I'm still planning to write a dedicated unit test for this, but for the sake of expediency, I wanted to provide at least one way for you to validate.

rockb1017 commented:
i figured out why. for MaxConcurrent, it was using 1. 
`cfg.MaxConcurrentFiles = 1`
now per batch, we open half of this number. so it wasn't able to open any file. I changed it to 2 and it works.
on this branch
```
pkg: github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file
BenchmarkFileInput/Single-16              504951              2003 ns/op             688 B/op         10 allocs/op
BenchmarkFileInput/Glob-16                122710              9888 ns/op            2761 B/op         42 allocs/op
BenchmarkFileInput/MultiGlob-16           123349              9809 ns/op            2761 B/op         42 allocs/op
BenchmarkFileInput/MaxConcurrent-16        65827             15271 ns/op            2757 B/op         42 allocs/op
BenchmarkFileInput/FngrPrntLarge-16       511382              2378 ns/op             690 B/op         10 allocs/op
BenchmarkFileInput/FngrPrntSmall-16       510836              2095 ns/op             688 B/op         10 allocs/op
```

on main
```
pkg: github.com/open-telemetry/opentelemetry-log-collection/operator/builtin/input/file
BenchmarkFileInput/Single-16              472940              2210 ns/op             688 B/op         10 allocs/op
BenchmarkFileInput/Glob-16                116560             12118 ns/op            2763 B/op         42 allocs/op
BenchmarkFileInput/MultiGlob-16           119990             13379 ns/op            2765 B/op         42 allocs/op
BenchmarkFileInput/MaxConcurrent-16        32445             30870 ns/op            2759 B/op         42 allocs/op
BenchmarkFileInput/FngrPrntLarge-16       508389              2413 ns/op             689 B/op         10 allocs/op
BenchmarkFileInput/FngrPrntSmall-16       474009              2588 ns/op             688 B/op         10 allocs/op

```

djaglowski commented:
Simple enough. However, we need to either support `max_concurrent_files=1`, or at least produce an error when validating the config if a user has specified `1`.

djaglowski commented on operator/builtin/input/file/util_test.go:
```diff
@@ -35,7 +35,7 @@ import (
 
 func newDefaultConfig(tempDir string) *InputConfig {
 	cfg := NewInputConfig("testfile")
-	cfg.PollInterval = helper.Duration{Duration: 50 * time.Millisecond}
```
Is this change necessary?

rockb1017 replied:
I think it being too short causes test to be flaky somehow. and 200ms is the default configuration anyway.

rockb1017 replied:
let me try reverting this.

djaglowski commented on operator/builtin/input/file/util_test.go:
```diff
@@ -122,7 +122,7 @@ func waitForOne(t *testing.T, c chan *entry.Entry) *entry.Entry {
 	case e := <-c:
 		return e
-	case <-time.After(time.Second):
+	case <-time.After(5 * time.Second):
```
Can we revert this (and the other occurrences)?

rockb1017 replied:
what's wrong with it?

rockb1017 replied:
I think it is necessary, maybe not all the way to 5 secs, but I am suspecting the windows server on github action is not fast enough. so when it is executing all the tests in parallel, it fails to complete the poll cycle.

djaglowski commented on docs/operators/file_input.md:
```diff
@@ -19,7 +19,7 @@ The `file_input` operator reads logs from files. It will place the lines read in
 | `fingerprint_size`     | `1kb`            | The number of bytes with which to identify a file. The first bytes in the file are used as the fingerprint. Decreasing this value at any point will cause existing fingerprints to forgotten, meaning that all files will be read from the beginning (one time). |
 | `max_log_size`         | `1MiB`           | The maximum size of a log entry to read before failing. Protects against reading large amounts of data into memory |
-| `max_concurrent_files` | 1024             | The maximum number of log files from which logs will be read concurrently. If the number of files matched in the `include` pattern exceeds this number, then files will be processed in batches. One batch will be processed per `poll_interval`. |
+| `max_concurrent_files` | 1024             | The maximum number of log files from which logs will be read concurrently. If the number of files matched in the `include` pattern exceeds half of this number, then files will be processed in batches. One batch will be processed per `poll_interval`. |
```
We need to either enforce and document a minimum of 2, or rework how `maxBatchSize` works, so that it can allocate the exact count of `max_concurrent_files`.

rockb1017 replied:
i would go with enforcing minimum of 2.

rockb1017 commented:
> Simple enough. However, we need to either support `max_concurrent_files=1`, or at least produce an error when validating the config if a user has specified `1`.

I think user putting 1 for max concurrent file is not very common use case. I will produce error for config validation

djaglowski reviewed (approved):
Thanks for your persistence @rockb1017!

LGTM

tigrannajaryan commented:
@djaglowski @rockb1017 judging from how much effort you needed to put into this it was not a trivial problem. Can we document how this works now? I do not feel that the code is sufficient for understanding. Even a retroactive design document with a review in Log SIG would not be bad to do (but maybe that's overkill and it is not that complicated, you tell me).

tigrannajaryan commented:
Meta note: for the future let's make sure we have proper PR title, commit title and commit message.

djaglowski commented:
@tigrannajaryan, I'll document how it currently works and how this change would work. The general design is not particularly complicated but the code could certainly be made more readable. Some small steps have been taken toward this end. More will be needed.

The primary challenge with this overall effort has been in establishing confidence in the edge cases, particularly where file rotation is involved. Several tests are probabilistic because they rely on the timing of system calls. We need to establish that these tests can be made deterministic without sidestepping some critical feature of the edge case. I think it will be most helpful to articulate a couple specific test scenarios and attempt to establish how they can be deterministically tested. 

I can spend some more time on this effort next week, and will share particulars with you. Perhaps we can together establish the best path forward from there.

tigrannajaryan commented:
@djaglowski sounds good, as soon as we are confident that the feature is rock solid and it is documented well enough for the future generations we are good.
